helix-view/src/document.rs

```diff
@@ -550,6 +555,112 @@ impl Document {
+                    let (tag, range_old, range_new) = op.as_tag_tuple();
+                    let (start, end) = {
+                        let origin = old.as_ptr();
+                        let words = &old_words[range_old];
+                        let start = words[0];
+                        let end = *words.last().unwrap();
+
+                        // Always safe because the slices were already created safely.
+                        let start_offset = unsafe { start.as_ptr().offset_from(origin) } as usize;
+                        let between_offset = unsafe { end.as_ptr().offset_from(start.as_ptr()) }
+                            as usize
+                            + end.len();
+
+                        // We need to convert from byte indices to char indices since that's what
+                        // `ChangeSet` operates on.
+                        (
+                            self.text().byte_to_char(start_offset),
+                            self.text().byte_to_char(start_offset + between_offset),
+                        )
+                    };
+
+                    match tag {
+                        similar::DiffTag::Insert | similar::DiffTag::Replace => {
+                            let words = &new_words[range_new];
+                            // Because `words` is a slice of slices, we need to concat them
+                            // back into one `&str`. We're using `unsafe` to avoid an allocation
+                            // that would happen if we were using the `std::slice::concat()`.
+                            let text: &str = {
+                                let origin = new.as_ptr();
+                                let start = words[0];
+                                let end = *words.last().unwrap();
+
+                                // Always safe because the slices were already created safely.
+                                let start_offset =
+                                    unsafe { start.as_ptr().offset_from(origin) } as usize;
+                                let between_offset =
+                                    unsafe { end.as_ptr().offset_from(start.as_ptr()) } as usize
+                                        + end.len();
+                                &new[start_offset..start_offset + between_offset]
+                            };
+
+                            Some((start, end, Some(text.into())))
+                        }
+                        similar::DiffTag::Delete => Some((start, end, None)),
+                        similar::DiffTag::Equal => None,
+                    }
+                })
+                .collect();
+
+            let transaction = Transaction::change(self.text(), changes.into_iter());
+            self.apply(&transaction, view_id);
+            self.append_changes_to_history(view_id);
+        } else {
+            return Err(anyhow::anyhow!("can't find file to reload from"));
+        }
```

Comment:
Using an early return with `if !path.exists()` at the beginning of the block would make it more readable (and remove an indent for the whole block)